odxtools/udsbinner.py

```diff
@@ -0,0 +1,123 @@
+# SPDX-License-Identifier: MIT
+from typing import Dict, Iterable, Optional
+
+from .diagservice import DiagService
+from .nameditemlist import NamedItemList
+from .parameters.codedconstparameter import CodedConstParameter
+
+
+class UdsBinner:
+    """Class to categorize a long list of services into the service
+    groups defined by the UDS standard.
+
+    This class is supposed to be used like this:
+
+    db = odxtools.load_file("my_cool_diagnostics_db.pdx")
+    ...
+    uds_bins = UdsBinner(db.ecus.my_ecu)
+    uds_bins.print_info()
+    """
+
+    def __init__(self, services: Iterable[DiagService]):
+        service_groups: Dict[Optional[int], NamedItemList[DiagService]] = {}
+        for service in services:
+            SID = self.__extract_sid(service)
+
+            if SID not in service_groups:
+                service_groups[SID] = NamedItemList()
+
+            service_groups[SID].append(service)
+
+        self.service_groups = service_groups
+
+    def __extract_sid(self, service: DiagService) -> Optional[int]:
+        # diagnostic services without requests are possible; just like
+        # aircraft without wings...
+        if service.request is None:
+            return None
+
+        prefix = 0  # prefix of constant parameters
+        cursor = 0  # bit position of the next parameter
+        for param in service.request.parameters:
+            if not isinstance(param, CodedConstParameter):
+                # we *need* at least the first byte of a request to be statically defined!
+                return None
+
+            param_len = param.get_static_bit_length()
+            if param_len is None:
+                return None
+
+            if not isinstance(param.coded_value, int):
+                return None
+
+            prefix <<= param_len
+            prefix |= param.coded_value & ((1 << param_len) - 1)
+            cursor += param_len
+
+            if cursor >= 8:
+                # we have a prefix that is at least 8 bits
+                # long. return its most significant byte.
+                prefix >>= cursor - 8
+                return prefix & 0xff
+
+        return None
+
+    def print_info(self) -> None:
```

Comment:
return the string, let user do the printing
```suggestion
    def __str__(self) -> None:
```

Reply:
great suggestion! implemented: 29a305a4c7c12757381e91938c8c6bf6e199e5f1